Enhance bdk chain structures

vladimirfomene opened the pull request:
### Description

Fixes bitcoindevkit/bdk#1061 

### Changelog notice

- Move WalletUpdate to the wallet module
- Remove ForEachTxout trait completely
- Refactor ElectrumExt to not use WalletUpdate.

### Checklists

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

vladimirfomene committed:
refactor: Edit ElectrumExt not to use WalletUpdate

vladimirfomene committed:
refactor: Remove ForEachTxout trait

vladimirfomene committed:
refactor: Move WalletUpdate to wallet module

vladimirfomene committed:
refactor: move WalletChangeset to wallet module

Consequently, remove the `WalletChangeset` dependency from
`example_electrum` and `example_esplora` examples.

LLFourn commented:
What's the motivation for removing `ForEachTxOut`? I think it serves a purpose in that it allows you to scan several things that are bunches of txouts (including things that are not transactions).

evanlinjin commented:
> What's the motivation for removing `ForEachTxOut`? I think it serves a purpose in that it allows you to scan several things that are bunches of txouts (including things that are not transactions).

This was my suggestion and I wrote it on the ticket. I don't think it is needed? The `Indexer` trait has the method `Indexer::index_txout(&mut self, outpoint: OutPoint, txout: &TxOut) -> Self::ChangeSet`, and we can have another method:

```rust
pub trait Indexer {
    /* ... */

    fn index_txouts<'a>(
        &mut self,
        txouts: impl IntoIterator<Item = (OutPoint, &'a TxOut)>,
    ) -> Self::ChangeSet
    where
        Self::ChangeSet: Append + Default,
    {
        let mut changeset = Self::ChangeSet::default();
        for (outpoint, txout) in txouts {
            Append::append(&mut changeset, self.index_txout(outpoint, txout));
        }
        changeset
    }
}
```

Now anything that implements `IntoIterator<Item = (OutPoint, &TxOut)>` will work.

vladimirfomene commented:
Will rebase this once MSRV fixes make it to master.

evanlinjin commented on crates/bdk/src/wallet/mod.rs:
```diff
@@ -94,6 +94,91 @@ pub struct Wallet<D = ()> {
+    /// Update for the [`LocalChain`].
+    ///
+    /// [`LocalChain`]: local_chain::LocalChain
+    pub chain: local_chain::Update,
```
@vladimirfomene Another request, can you try wrap this with an `Option` please?

I think it'll be important to not have to update with a `CheckPoint`.

evanlinjin commented on crates/chain/src/keychain/txout_index.rs:
```diff
@@ -124,13 +124,11 @@ impl<K: Clone + Ord + Debug> KeychainTxOutIndex<K> {
+            changeset.append(self.scan_txout(OutPoint::new(tx.txid(), op as u32), txout));
+        }
         changeset
     }
```
I think we can move `scan()` to the `Indexer::index_tx()` implementation of `KeychainTxOutIndex`. `scan_txout` can also be replaced by `Indexer::index_txout`.

I would also do the same for `SpkTxoutIndex`.

LLFourn replied:
I think this is a good idea. Also `index_tx` can have a default impl in the trait.

evanlinjin replied:
Created ticket: #1097

vladimirfomene committed:
refactor: Allow for no chain update

evanlinjin commented on crates/bdk/src/wallet/mod.rs:
```diff
@@ -94,6 +94,91 @@ pub struct Wallet<D = ()> {
     secp: SecpCtx,
 }
 
+/// A structure to update [`KeychainTxOutIndex`], [`TxGraph`] and [`LocalChain`] atomically.
```
We will need to change this to something more understandable. This is no longer the `bdk_chain` crate.

```rust
/// A structure to update [`Wallet`].
```

vladimirfomene replied:
I'm thinking we can have something like:
```
/// A structure to update [`Wallet`].

/// It updates [`bdk_chain::KeychainTxoutIndex`], [`bdk_chain::TxGraph`] and [`bdk_chain::LocalChain`] atomically.
```

evanlinjin commented on crates/bdk/src/wallet/mod.rs:
```diff
@@ -94,6 +94,91 @@ pub struct Wallet<D = ()> {
+            graph: TxGraph::default(),
+            chain: Some(chain_update),
+        }
+    }
```
Since the `chain` field is now an `Option`, we can implement `Default` for `WalletUpdate` instead of this method.

Don't use `derive()`, otherwise we will need to have `K` and `A` implement `Default` for `WalletUpdate` to be `Default`.

vladimirfomene replied:
Let me do this

evanlinjin commented on crates/electrum/src/electrum_ext.rs:
```diff
@@ -539,9 +526,8 @@ fn populate_with_spks<K, I: Ord + Clone>(
             }
 
             for tx in spk_history {
-                let tx_entry = update.graph_update.entry(tx.tx_hash).or_default();
```
I think you have accidentally removed the ability for us to fetch unconfirmed/mempool transactions.

Notice the `or_default()` call here.

evanlinjin commented on crates/electrum/src/electrum_ext.rs:
```diff
@@ -1,29 +1,28 @@
 #[derive(Debug, Clone)]
 pub struct ElectrumUpdate<K, A> {
-    /// Map of [`Txid`]s to associated [`Anchor`]s.
-    pub graph_update: HashMap<Txid, BTreeSet<A>>,
```
What is the motivation of this change?

vladimirfomene replied:
The motivation was to get rid of `WalletUpdate` but still have a way to get all the graph updates that we were exposing through `WalletUpdate`.

LLFourn replied:
The main motivation for removing `WalletUpdate` from here is so `keychain_update` doesn't have to be part of this struct since it is not always necessary to have it. The one that does the keychain scan should just return a tuple with the keychain update being an element. This removes a type parameter. The `new_tip` also seems orthogonal -- it can just be returned as another element in the tuple. So `ElectrumUpdate` should only have the graph info in it. 

The particular change made here seems a bit dangerous since we can't enforce the user doesn't use the `TxGraph` as an update before it's finalized. If we do what i suggest above I think it's fine to leave this as a `HashMap` and then consume self in the `finalize` methods.

evanlinjin reviewed (changes requested):
Thanks for the work so far. Sorry about my incremental reviews.

LLFourn commented:
> > What's the motivation for removing `ForEachTxOut`? I think it serves a purpose in that it allows you to scan several things that are bunches of txouts (including things that are not transactions).
> 
> This was my suggestion and I wrote it on the ticket. I don't think it is needed? The `Indexer` trait has the method `Indexer::index_txout(&mut self, outpoint: OutPoint, txout: &TxOut) -> Self::ChangeSet`, and we can have another method:

So if I want to index a `Block` what should I do?. I can iterate manually over every `TxOut` and call this but that seems much more cumbersome than it needs to be?

vladimirfomene commented:
@LLFourn I'm thinking you will use `index_tx` here instead of `index_txout`.

LLFourn commented on crates/electrum/src/electrum_ext.rs:
```diff
@@ -118,7 +108,7 @@ impl<K> ElectrumUpdate<K, ConfirmationHeightAnchor> {
 
         let graph_changeset = {
-            let old_changeset = TxGraph::default().apply_update(update.graph.clone());
+            let old_changeset = TxGraph::default().apply_update(self.graph_update);
```
Not part of this PR necessarily but the code below seems super complicated and uses changesets for no reason? How about we just have a function on `tx_graph.map_anchor(|anchor| get_new_anchor(anchor))` which consumes self and returns a new `TxGraph` with the new anchor.

evanlinjin replied:
Created issue: #1096

LLFourn commented on crates/electrum/src/electrum_ext.rs:
```diff
@@ -54,29 +55,19 @@ impl<K, A: Anchor> ElectrumUpdate<K, A> {
-            }
-            for anchor in anchors {
-                let _ = graph_update.insert_anchor(txid, anchor);
+                let _ = self.graph_update.insert_seen_at(tx.txid(), seen_at);
```
I think this is wrong but not due to this PR. We should only set `seen_at` if the transaction is not anchored. Right @evanlinjin?

evanlinjin replied:
I think it's okay to have both. The transaction may get reorged out and the last_seen will become useful

LLFourn commented:
> @LLFourn I'm thinking you will use `index_tx` here instead of `index_txout`.

Understood. I think might be as easy as we need to make it. Thanks.

vladimirfomene committed:
refactor: Implement Default for WalletUpdate

vladimirfomene committed:
refactor: Remove `scan` and `scan_txout` from SpkTxoutIndex and KeychainTxoutIndex

evanlinjin commented on crates/electrum/src/electrum_ext.rs:
```diff
@@ -58,7 +58,7 @@ impl<K, A: Anchor> ElectrumUpdate<K, A> {
         seen_at: Option<u64>,
         missing: Vec<Txid>,
-    ) -> Result<WalletUpdate<K, A>, Error> {
+    ) -> Result<(TxGraph<A>, BTreeMap<K, u32>, local_chain::CheckPoint), Error> {
```
Please return a `local_chain::Update` instead of `local_chain::CheckPoint`.

evanlinjin committed:
feat(electrum)!: change signature of `ElectrumExt`

We remove `ElectrumUpdate` and return tuples instead for `ElectrumExt`
methods. We introduce the `IncompleteTxGraph` structure to specifically
hodl the incomplete `TxGraph`.

This change is motivated by @LLFourn's comment: https://github.com/bitcoindevkit/bdk/pull/1084/commits/794bf37e637d3266b75827678f015e14f827d318#r1305432603

evanlinjin reviewed (approved):
ACK fb8777a9c3ecfec0aeca35dfbd773e51b0cf8e56

vladimirfomene commented:
Thanks @evanlinjin for these improvements!

notmandatory commented:
Per Evan moved this to Alpha.2 since it's required for RPC PR

notmandatory reviewed (approved):
ACK c56728ff1315e0deaf256af07fd1ff5e18fced8a

I confirmed `example_electrum` and `example_esplora` scan and sync commands still work. Used my test signet descriptors:

```shell
cargo run -- --network signet "wpkh(tprv8ZgxMBicQKsPdy6LMhUtFHAgpocR8GC6QmwMSFpZs7h6Eziw3SpThFfczTDh5rW2krkqffa11UpX3XkeTTB2FvzZKWXqPY54Y6Rq4AQ5R8L/84'/1'/0'/0/*)" "wpkh(tprv8ZgxMBicQKsPdy6LMhUtFHAgpocR8GC6QmwMSFpZs7h6Eziw3SpThFfczTDh5rW2krkqffa11UpX3XkeTTB2FvzZKWXqPY54Y6Rq4AQ5R8L/84'/1'/0'/1/*)" scan
```

notmandatory commented:
One small nit, it looks like most of the commits include breaking changes so commit messages should include a `!` to denote commit breaks old APIs.  Doesn't need to be fixed for this PR but to keep in mind for future.

evanlinjin commented:
ACK c56728ff1315e0deaf256af07fd1ff5e18fced8a

LLFourn committed:
feat: impl Append for lots of tuples

LLFourn commented on crates/bdk/src/wallet/mod.rs:
```diff
@@ -95,6 +95,90 @@ pub struct Wallet<D = ()> {
     secp: SecpCtx,
 }
 
+/// A structure to update [`Wallet`].
```
```suggestion
/// An update to [`Wallet`].
``` 

There is no need to refer to it as a structure. Anyone can see it is a `struct`.

LLFourn commented on crates/bdk/src/wallet/mod.rs:
```diff
@@ -95,6 +95,90 @@ pub struct Wallet<D = ()> {
+    /// [`KeychainTxOutIndex`].
+    pub last_active_indices: BTreeMap<K, u32>,
+
+    /// Update for the [`TxGraph`].
```
```suggestion
    /// Update for the wallet's internal [`TxGraph`].
```

LLFourn commented on crates/bdk/src/wallet/mod.rs:
```diff
@@ -95,6 +95,90 @@ pub struct Wallet<D = ()> {
+    /// Update for the [`TxGraph`].
+    pub graph: TxGraph<A>,
+
+    /// Update for the [`LocalChain`].
```
```suggestion
    /// Update for the wallet's internal [`LocalChain`].
```

LLFourn commented on crates/bdk/src/wallet/mod.rs:
```diff
@@ -95,6 +95,90 @@ pub struct Wallet<D = ()> {
+    }
+}
+
+/// A structure that records the corresponding changes as result of applying an [`WalletUpdate`].
```
```suggestion
/// The changes made to a wallet by applying an [`WalletUpdate`].
```

LLFourn commented on crates/bdk/src/wallet/mod.rs:
```diff
@@ -95,6 +95,90 @@ pub struct Wallet<D = ()> {
+        deserialize = "K: Ord + serde::Deserialize<'de>, A: Ord + serde::Deserialize<'de>",
+        serialize = "K: Ord + serde::Serialize, A: Ord + serde::Serialize",
+    ))]
+    pub index_tx_graph: indexed_tx_graph::ChangeSet<A, keychain::ChangeSet<K>>,
```
```suggestion
    pub indexed_tx_graph: indexed_tx_graph::ChangeSet<A, keychain::ChangeSet<K>>,
```

LLFourn replied:
I don't recall how this name change happened but it shouldn't have.

LLFourn commented on crates/bdk/src/wallet/mod.rs:
```diff
@@ -95,6 +95,90 @@ pub struct Wallet<D = ()> {
+}
+
 /// The update to a [`Wallet`] used in [`Wallet::apply_update`]. This is usually returned from blockchain data sources.
 pub type Update = WalletUpdate<KeychainKind, ConfirmationTimeAnchor>;
```
Remove these type Alias since they are no longer needed.

LLFourn commented on crates/bdk/src/wallet/mod.rs:
```diff
@@ -95,6 +95,90 @@ pub struct Wallet<D = ()> {
+///
+/// It updates [`bdk_chain::keychain::KeychainTxOutIndex`], [`bdk_chain::TxGraph`] and [`local_chain::LocalChain`] atomically.
+#[derive(Debug, Clone)]
+pub struct WalletUpdate<K, A> {
```
```suggestion
pub struct Update {
```

`K` and `A` are `KeychainKind` and `ConfirmationTimeAnchor` and are invariant at the moment.

LLFourn commented on crates/bdk/src/wallet/mod.rs:
```diff
@@ -95,6 +95,90 @@ pub struct Wallet<D = ()> {
+
+/// A structure that records the corresponding changes as result of applying an [`WalletUpdate`].
+#[derive(Debug, Clone, PartialEq, serde::Serialize, serde::Deserialize)]
+pub struct WalletChangeSet<K, A> {
```
```suggestion
pub struct ChangeSet {
```

Same reasoning as above.

LLFourn commented on crates/chain/src/spk_txout_index.rs:
```diff
@@ -54,19 +53,35 @@ impl<I> Default for SpkTxOutIndex<I> {
 
 impl<I: Clone + Ord> Indexer for SpkTxOutIndex<I> {
-    type ChangeSet = ();
+    type ChangeSet = BTreeSet<I>;
```
This doesn't make sense to me. The changeset is meant to keep note of an index's persisted state. But `SpkTxOutIndex` has no persisted state and what is returned from `index_txout` doesn't represent a change to state at all. This shoulds stay as `()`.

LLFourn commented on crates/electrum/src/electrum_ext.rs:
```diff
@@ -15,90 +14,67 @@ use std::{
+/// determine the full transactions missing from [`TxGraph`]. Then call [`Self::finalize`] to
+/// fetch the full transactions from Electrum and finalize the update.
+#[derive(Debug, Default, Clone)]
+pub struct IncompleteTxGraph<A>(HashMap<Txid, BTreeSet<A>>);
```
Can we call this just `RelevantTxids` there's no need to conceptually couple this to a `TxGraph`.

Let me try and point out what seems overly complicated here. I know these problems were not introduced in this PR but i think now's the time to fix them.

1. The data you get from electrum is always anchored with `ConfirmationHeight` so why is there a type parameter here.
2. Why is there a type parameter on `ElectrumExt`. Extension traits don't have type parameters in general.
3. These "finalize" methods should just be `into_tx_graph` and `into_confirmation_time_tx_graph`

LLFourn commented on example-crates/example_electrum/src/main.rs:
```diff
@@ -60,19 +60,22 @@ pub struct ScanOptions {
+);
 
 fn main() -> anyhow::Result<()> {
     let (args, keymap, index, db, init_changeset) =
```
:bulb: deconstruct `init_changeset` into `(disk_local_chain, disk_tx_graph)` to avoid using `.1` and `.0` below.

LLFourn commented on example-crates/wallet_electrum/src/main.rs:
```diff
@@ -52,14 +53,21 @@ fn main() -> Result<(), Box<dyn std::error::Error>> {
         .collect();
 
-    let electrum_update = client.scan(prev_tip, keychain_spks, None, None, STOP_GAP, BATCH_SIZE)?;
+    let (chain_update, incomplete_graph_update, keychain_update) =
```
```suggestion
    let (chain_update, relevant_txids, keychain_update) =
```

LLFourn commented on crates/chain/src/indexed_tx_graph.rs:
```diff
@@ -233,7 +233,18 @@ pub trait Indexer {
+    ///
+    /// If the matched script pubkey is part of the lookahead, the last stored index is updated for
+    /// the script pubkey's keychain and the [`ChangeSet`] returned will reflect the
+    /// change.
```
lookahead has nothing to do with this trait. That's an API of `KeychainTxOutIndex`. I think this documentation is missing the point (probably because it was copy pasted) for this methid. The documentation of the trait should mention that it's primarily used by `IndexedTxGraph`. Then we can explain here that `index_tx` is called when there's a new transaction added to the graph.

LLFourn commented on crates/chain/src/keychain/txout_index.rs:
```diff
@@ -200,14 +176,11 @@ impl<K: Clone + Ord + Debug> KeychainTxOutIndex<K> {
     /// Set the lookahead count for `keychain`.
     ///
     /// The lookahead is the number of scripts to cache ahead of the last stored script index. This
```
```suggestion
    /// The lookahead is the number of scripts to cache ahead of the last revealed script index. This
```

LLFourn commented on crates/chain/src/keychain/txout_index.rs:
```diff
@@ -200,14 +176,11 @@ impl<K: Clone + Ord + Debug> KeychainTxOutIndex<K> {
     ///
     /// The lookahead is the number of scripts to cache ahead of the last stored script index. This
-    /// is useful during a scan via [`scan`] or [`scan_txout`].
+    /// is useful during a scan via [`Indexer::index_tx`] or [`Indexer::index_txout`].
```
Since we are not necessarily encouraging people to "scan" things with these methods (it's mostly done automatically) I don't think this sentence is doing much for us. Instead explain when it is useful at a higher level i.e. when you are processing block data and want to find outputs you own beyond your last revealed index because the last revealed index may not be accurate (some other system may have revealed script pubkeys beyond it).

LLFourn commented on crates/chain/src/lib.rs:
```diff
@@ -100,3 +100,11 @@ pub mod collections {
 /// How many confirmations are needed f or a coinbase output to be spent.
 pub const COINBASE_MATURITY: u32 = 100;
+
+impl<A, IA> From<indexed_tx_graph::ChangeSet<A, IA>>
```
Not sure if we need this. If we do it shouldn't be here (it seems to be about `indexed_tx_graph`)

LLFourn commented on crates/electrum/src/electrum_ext.rs:
```diff
@@ -223,9 +204,8 @@ impl ElectrumExt<ConfirmationHeightAnchor> for Client {
-            let mut update = ElectrumUpdate::<K, ConfirmationHeightAnchor>::new(tip.clone());
-            let cps = update
-                .new_tip
+            let mut graph_update = IncompleteTxGraph::<ConfirmationHeightAnchor>::default();
```
In general can we avoid the term "graph" in this code (unless we are actually returning a `TxGraph`). We are just finding relevant transactions to set of script pubkeys.

LLFourn commented on crates/electrum/src/electrum_ext.rs:
```diff
@@ -165,6 +136,7 @@ pub trait ElectrumExt<A> {
     /// The scan for each keychain stops after a gap of `stop_gap` script pubkeys with no associated
     /// transactions. `batch_size` specifies the max number of script pubkeys to request for in a
     /// single batch request.
+    #[allow(clippy::type_complexity)]
```
Hmm I don't think we should do this. Can we just package these tuple types into their own types. Consider bundling (`local_chain::Update, RelevantTxids)` into its own type.

LLFourn reviewed (changes requested):
Hey @vladimirfomene thanks for doing this refactoring.

I've pointed out a bunch of ways we can simplify things in both wallet types and electrum_ext. Many of the changes requested are not to do with this PR but I think this is the place to do them.

[edit] I also pushed 41042069809e3eeb4a8cc8a5a8db1af2c57c4a11

vladimirfomene committed:
refactor: improve docs, cleanup unnecessary types and improve code

evanlinjin commented on crates/chain/src/keychain/txout_index.rs:
```diff
@@ -120,6 +112,36 @@ impl<K: Clone + Ord + Debug> Indexer for KeychainTxOutIndex<K> {
+    /// your txouts.
+    /// 2. When getting new data from the chain, you usually scan it before incorporating it into
+    /// your chain state.
+    pub fn scan(&mut self, tx: &bitcoin::Transaction) -> super::ChangeSet<K> {
```
Why are we reintroducing `scan` and `scan_txout`? The `Indexer` methods already do the same thing.

Same question for `SpkTxOutIndex`.

vladimirfomene replied:
I can see how to get rid of `scan` and `scan_txout` in `TxoutIndex` but I don't see how to get rid of it in `SpkTxoutIndex` as `scan_txout` in `TxoutIndex` depends on `scan_txout` in `SpkTxoutIndex` returning an optional index. Do you mind sharing some code sample of how this could work?

evanlinjin replied:
Thanks for explaining, I see the problem now.

In `KeychainTxOutIndex`, the signature of the methods are exactly the same so we don't need to duplicate them (rm `scam` and `scan_txout`). For `SpkTxOutIndex`, they are different, so we need different methods.

LLFourn commented on crates/chain/src/keychain/txout_index.rs:
```diff
@@ -199,15 +174,10 @@ impl<K: Clone + Ord + Debug> KeychainTxOutIndex<K> {
-    /// [`scan_txout`]: Self::scan_txout
+    /// The lookahead is the number of scripts to cache ahead of the last revealed script index. This
+    /// is useful to find outputs you own when processing block data that lie beyond the last revealed
+    /// index. This is because in certain scenarios we might not know the last revealed indexit in like
```
Thanks this explanation is better. There's a typo in this sentence. I think it just needs a small rewrite.

PS does this method still panic if the keychain doesn't exist? If so we should probably still mention that.

vladimirfomene replied:
Let me check. If it still panics, I will mention it.

LLFourn commented on crates/electrum/src/electrum_ext.rs:
```diff
@@ -140,21 +114,17 @@ impl<K> ElectrumUpdate<K, ConfirmationHeightAnchor> {
     }
 }
 
+type ElectrumUpdate = (local_chain::Update, RelevantTxids);
```
I think it would be better if this was a struct and explain why the update is in this form (we have to update the chain and the txids at the same time since we anchor the txids to the same chain tip that we check before and after we gather the txids).

LLFourn reviewed (approved):
This LGTM! ACK ee01b0fd632c1ff4ab95ec6a50e6af61a1eab7df

left a couple of small comments.

evanlinjin commented on crates/chain/src/indexed_tx_graph.rs:
```diff
@@ -225,15 +225,21 @@ impl<A, K> From<keychain::ChangeSet<K>> for ChangeSet<A, keychain::ChangeSet<K>>
+    }
+}
+
+/// Defines methods to index transaction data.
```
```suggestion
/// A trait for indexing transaction data.
///
/// Types which implement this trait can be used to construct an [`IndexedTxGraph`]. This trait
/// should rarely be called directly.
```

vladimirfomene replied:
What do you mean by `This trait should rarely be called directly`?

evanlinjin replied:
The caller shouldn't be calling the trait methods directly. It should all be handled in `IndexedTxGraph`.

---

An example of stdlib documentation using similar wording: https://doc.rust-lang.org/stable/std/iter/trait.Product.html

evanlinjin commented on crates/chain/src/indexed_tx_graph.rs:
```diff
@@ -225,15 +225,21 @@ impl<A, K> From<keychain::ChangeSet<K>> for ChangeSet<A, keychain::ChangeSet<K>>
+    fn from(indexed_changeset: ChangeSet<A, IA>) -> Self {
+        (local_chain::ChangeSet::default(), indexed_changeset)
+    }
+}
```
This is rather odd. Where is this useful?

vladimirfomene replied:
It is useful in `example_cli::handle_commands`. If not implemented we get the following error, `the trait From<bdk_chain::indexed_tx_graph::ChangeSet<ConfirmationTimeAnchor, bdk_chain::keychain::ChangeSet<Keychain>>> is not implemented for (BTreeMap<u32, std::option::Option<BlockHash>>, bdk_chain::indexed_tx_graph::ChangeSet<ConfirmationTimeAnchor, bdk_chain::keychain::ChangeSet<Keychain>>)`

evanlinjin replied:
@vladimirfomene surely there must be a way around this?

vladimirfomene replied:
Let me look for another solution

evanlinjin replied:
Thanks, let me know what you find.

evanlinjin commented on crates/chain/src/spk_txout_index.rs:
```diff
@@ -77,49 +76,37 @@ impl<I: Clone + Ord> Indexer for SpkTxOutIndex<I> {
-
 impl<I: Clone + Ord> SpkTxOutIndex<I> {
-    /// Scans an object containing many txouts.
+    /// Scans a transaction for txouts.
```
```suggestion
    /// Scans a transaction's outputs for matching script pubkeys.
```

evanlinjin commented on crates/chain/src/spk_txout_index.rs:
```diff
@@ -77,49 +76,37 @@ impl<I: Clone + Ord> Indexer for SpkTxOutIndex<I> {
-        txouts.for_each_txout(|(op, txout)| {
-            if let Some(spk_i) = scan_txout!(self, op, txout) {
+        for (i, txout) in tx.output.iter().enumerate() {
+            let op = OutPoint::new(tx.txid(), i as u32);
```
**Nit:** I think it would be better to create the `txid` outside the loop to avoid rehashing the transaction in every iteration.

vladimirfomene replied:
I agree

evanlinjin reviewed (changes requested):
Mostly nits and documentation suggestions.

evanlinjin committed:
fix(chain)!: rm weird `From` impl

And signature of `example_cli::KeychainChangeSet` is changed.

evanlinjin reviewed (approved):
ACK 1ff806c67f4da9ba58b7c7689fde0fe41a34a6f5
